# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](https://www.datacamp.com/community/tutorials/tutorial-jupyter-notebook##UseJupyter).

Additionally, in order to write text in this document, you will be making use of Markdown cells. A handy cheatsheet for Markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1

Note: the exercise numbers below match up with the reading for ease of cross referencing.

### Exercise 1.4

Start the Python interpreter and use it as a simple calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *

* How many seconds are there in 42 minutes 42 seconds?
* How many miles are there in 10 kilometers? Hint: there are 1.61 kilometers in a mile.
* If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average pace (time per mile in minutes and seconds)? What is your average speed in miles per hour?

Please use the following markdown cell to provide your answers.

ran 6.21 miles in 2562 seconds. Average pace: 2562/6.21. Average speed: 1mi/6.876min * 60min/1hr.

In [41]:
(42 * 60) + 42     #seconds in 42 min 42 sec

2562

In [42]:
10 / 1.61      #miles in 10 km

6.211180124223602

In [43]:
(2562 / 6.21) / 60      #sec / mi

6.876006441223833

In [44]:
(1 / 6.876) * (60)     #average speed: mi / hr

8.726003490401396

Average pace: 6 minutes 48 seconds per mile

Average speed: 8.73 miles per hour

## Chapter 2

### Exercise

Assume that we execute the following assignment statements:

```
width = 17
height = 12.0
delimiter = '.'
```

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

Example: `width * 2` <br/>
Answer: 34, integer

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

In [45]:
width = 17
height = 12.0
delimiter = '.'

print(width/2)     #float
print(width/2.0)   #float
print(height/3)    #float
print(1 + 2 * 5)   #integer
delimiter * 5      #string

8.5
8.5
4.0
11


'.....'

### Exercise 2.2

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

The volume of a sphere with radius r is $4/3 \pi  r^3$. What is the volume of a sphere with radius 5? <br />
Hint: 392.7 is wrong! You can approximate pi with 3.14 or import it through the Python math library.

- In python2 they would've rounded 4/3 to 1. The way you would've fixed it is to make 4 or 3 a "float" (ie 4.0 or 3.0) and then it wouldn't have approximated anymore because it would no longer be dividing between 2 integers.

In [46]:
(4/3)*3.14 * (5**3)

523.3333333333334

Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [47]:
(24.95 * 0.6 *60) + 3 + 0.75*59 #float in dollars

945.4499999999999

If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [48]:
print(2*(8*60)+15 + 3*(7*60)+12)  #seconds to run all 2 easy pace + 3 tempo miles
print(2247 / 60)                  #minutes
print(37.45 - 8)                  #minutes past 7am

2247
37.45
29.450000000000003


Home for breakfast around 7:29am!

## Chapter 3

### Exercise 3.1

Write a function named `right_justify` that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('monty')
                                                                 monty
```
Hint: Use string concatenation and repetition. Also, Python provides a built-in function called `len` that returns the length of a string, so the value of `len('monty')` is 5.

Format method: 

In [49]:
def right_justify(s):
    print("{:>70}".format(s))
    
right_justify("monty")

print(len('monty'))

                                                                 monty
5


### Exercise 3.3

Note: This exercise should be done using only the statements and other features we have learned so far.

1\. Write a function that draws a grid like the following:

```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
Note: There are four empty spaces between the vertical bars.

To print more than one value on a line, you can print a comma-separated sequence of values:

```
print('+', '-')
```

By default, `print` advances to the next line, but you can override that behavior and put a space at the end, like this:

```
print('+', end=' ')
print('-')
```

The output of these statements is '+ -'.

- A `print` statement with no argument ends the current line and goes to the next line.

In [50]:
def stuff(a):
    x = '|         '+'|         '+'|'
    y = '+ '+'- '*4+'+ '+'- '*4+'+'
    print(y)
    print(x+'\n'+x+'\n'+x+'\n'+x)
    print(y)
    print(x+'\n'+x+'\n'+x+'\n'+x)
    print(y)
stuff('hello')

+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +


2\. Write a function that draws a similar grid with four rows and four columns.

In [51]:
def stuff(a):
    x = '|         '+'|         '+'|         '+'|'
    y = '+ '+'- '*4+'+ '+'- '*4+'+ '+'- '*4+'+'
    print(y)
    print(x+'\n'+x+'\n'+x+'\n'+x)
    print(y)
    print(x+'\n'+x+'\n'+x+'\n'+x)
    print(y)
    print(x+'\n'+x+'\n'+x+'\n'+x)
    print(y)
stuff('bye')

+ - - - - + - - - - + - - - - +
|         |         |         |
|         |         |         |
|         |         |         |
|         |         |         |
+ - - - - + - - - - + - - - - +
|         |         |         |
|         |         |         |
|         |         |         |
|         |         |         |
+ - - - - + - - - - + - - - - +
|         |         |         |
|         |         |         |
|         |         |         |
|         |         |         |
+ - - - - + - - - - + - - - - +


## Chapter 5.1 - 5.7

### Exercise 5.2

Fermat’s Last Theorem says that there are no positive integers $a$, $b$, and $c$ such that $$a^n + b^n = c^n$$ for any values of $n$ greater than 2.

1\. Write a function named `check_fermat` that takes four parameters – $a$, $b$, $c$ and $n$ – and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$, then the program should print: "Holy smokes, Fermat was wrong!" Otherwise the program should print: "No, that doesn’t work."

In [52]:
def check_fermat(a,b,c,n):
    a>0
    b>0
    c>0
    n>2
    if a**n + b**n == c**n :
        print('Holy smokes, Fermat was wrong!')
    else: print('No, that doesn\'t work.')
check_fermat(1,1,1,3)

No, that doesn't work.


2\. Write a function that prompts the user to input values for $a$, $b$, $c$ and $n$, converts them to integers, and uses `check_fermat` to check whether they violate Fermat’s theorem.

In [53]:
def fermat_part2(a,b,c,n):
    a = int('What is a:')

### Exercise 5.3 

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

1\. Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [54]:
def is_triangle(a,b,c):
    if a>b+c: print('No')
    elif b>a+c: print('No')
    elif c>a+b: print('No')
    else: print('Yes')
is_triangle(2,2,4)

Yes


2\. Write a function that prompts the user to input three stick lengths, converts them to integers, and uses `is_triangle` to check whether sticks with the given lengths can form a triangle.

In [55]:
def triangle_test(a,b,c):
    is_triangle

## Quick poll
About how long did you spend working on this Reading Journal?

1:10  +  1:35  +  0:45  =  3.5 hours

## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.

I didn't get a chance to ask a NINJA about part 2 of Exercises 5. 2 and 5.3 but I found them a bit confusing and could use a walk-through. (i.e. The question that says, "Write a function that prompts the user to input values for [variables used in part 1], converts them to integers, and uses [function written in part 1] to check what outcome they produce.) 

Overall though, it's going well. So far I'm actually enjoying the "Reading Journal" format. Thanks for soliciting feedback - it helps me reflect on my progress and figure out what I need to move forward!